### ISSUE TITLE
Additional Input Planes

### ISSUE THREAD

**Issue by AlexisOlson** [LeelaChessZero/lc0#1258]
Not really a request so much as a place to save ideas...

## Current input planes:
(6 piece types x 2 colors + isRep ) x 8 plies + 4 castling + isBlackMove + 50moveRuleCounter

![image](https://user-images.githubusercontent.com/10537957/80769867-2aafea00-8b14-11ea-87f6-ec4679f3d28e.png)

Source: https://slides.com/crem/lc0#/14

## Possible additional input planes:

![image](https://user-images.githubusercontent.com/10537957/80773651-22f64280-8b20-11ea-97c8-cfbece73d9fe.png)

For each piece set (examples shown for white N, P, and Q for the board above).

Gray cells are for reference only. They are not part of the plane.

### Motion Plane
All possible move squares with an empty board

![MotionN](https://user-images.githubusercontent.com/10537957/80772817-90ed3a80-8b1d-11ea-8863-a26e51c94bea.png) . ![MotionP](https://user-images.githubusercontent.com/10537957/80774341-5d60df00-8b22-11ea-84b3-66682e5fbc03.png) . ![MotionQ](https://user-images.githubusercontent.com/10537957/80772827-96e31b80-8b1d-11ea-976c-99df561f840b.png)

### Move Plane
All possible legal move squares

![MoveN](https://user-images.githubusercontent.com/10537957/80772876-c1cd6f80-8b1d-11ea-9273-8f6df957075f.png) . ![MoveP](https://user-images.githubusercontent.com/10537957/80774355-6baefb00-8b22-11ea-82a8-0987aba3227d.png) . ![MoveQ](https://user-images.githubusercontent.com/10537957/80772887-c8f47d80-8b1d-11ea-8a0f-68fb7c680716.png)

### Attack Plane
Opponent squares that can be captured

![AttackN](https://user-images.githubusercontent.com/10537957/80772898-d3af1280-8b1d-11ea-9e72-e330de056b09.png) . ![AttackP](https://user-images.githubusercontent.com/10537957/80774369-75386300-8b22-11ea-84b8-1bc1e3072965.png) . ![AttackQ](https://user-images.githubusercontent.com/10537957/80772904-d9a4f380-8b1d-11ea-8858-df32a7011c27.png)

### Threaten plane
Opponent squares that can capture the current piece

![ThreatenN](https://user-images.githubusercontent.com/10537957/80772920-e590b580-8b1d-11ea-8f1a-610eca45a8ba.png) . ![ThreatenP](https://user-images.githubusercontent.com/10537957/80774379-7f5a6180-8b22-11ea-92fd-1e26a964a1e2.png) . ![ThreatenQ](https://user-images.githubusercontent.com/10537957/80772927-ea556980-8b1d-11ea-8aa7-35db7d7b714c.png)

### Defend Plane
Squares with own pieces covered by the current piece

![DefendN](https://user-images.githubusercontent.com/10537957/80772939-f6d9c200-8b1d-11ea-8959-7575ade1e0bf.png) . ![DefendP](https://user-images.githubusercontent.com/10537957/80774395-884b3300-8b22-11ea-89a6-c8f92ddd9f78.png) . ![DefendQ](https://user-images.githubusercontent.com/10537957/80772947-fd683980-8b1d-11ea-8ba7-45abcfc94993.png)

### Shield Plane
Squares with own pieces blocked by the current piece

![ShieldN](https://user-images.githubusercontent.com/10537957/80772958-0c4eec00-8b1e-11ea-82f7-0919c05ece5d.png) . ![ShieldP](https://user-images.githubusercontent.com/10537957/80774412-91d49b00-8b22-11ea-9536-ebc90a948770.png) . ![SheildQ](https://user-images.githubusercontent.com/10537957/80772962-0f49dc80-8b1e-11ea-9982-505eed075913.png)

### Color Planes
Light and/or dark squares

![Color1](https://user-images.githubusercontent.com/10537957/80773023-3ef8e480-8b1e-11ea-815c-24fc13c681d2.png) . ![Color2](https://user-images.githubusercontent.com/10537957/80773032-44562f00-8b1e-11ea-984d-e59e3879c8b6.png)

Bonus: Maybe separate light and dark bishops as different pieces since they never transpose.

Previously mentioned on Discord:
https://discordapp.com/channels/425419482568196106/425419999096733706/702886783930728599

**Comment by TesseractA** [comment]
https://github.com/LeelaChessZero/lc0/issues/882

https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/lczero/350gHBeUUiQ/2qtoWJ92AAAJ

Also, we've tried Bishop Color planes--Error323 did it. I can't remember exactly what those tests.

**Comment by TesseractA** [comment]
Posting here from #882 and closing the other issue to prevent redundancies.
```The idea is to give the neural network separate binary masks representing each possible move by each piece. The theory is that it will improve tactical recognition and improve pin/skewer detection. The problem is, there are several different implementations of the same idea and I'm having trouble determining which ideas are best:

-32-64 for each piece/pawn on the board. (When there are extra pieces/pawns on the board for variants, that could potentially run into issues with only 32 pieces.)

-18 rook, 18 bishop, 18 queen attack planes (forgoing the king, knight, and pawn attack planes for simplicity) And maybe ignore putting in extra attack planes for bishops, queens, rooks, and/or knights? We may never reach a real position with 8 queens on both sides anyway. For any such cases, we build in exceptions that say to use a neural network that is either compatible with that many pieces or we forgo putting in some of the pieces.

-Randomly putting pieces into different attack planes of the same or different types (rooks could be on the same planes as bishops are, this would reduce the amount of planes needed to 28 or less.)

-Fixed attack planes for each piece. (Could possibly cause some board symmetry problems)

-Do we preserve and reserve attack masks for a single side or allow attack masks to be used from both sides on the same input plane?

I'd also like other thoughts, regardless of whether they are about this being 'zero' or not.

The attack binary mask board generation should be easy and any necessary infrastructure could be borrowed from Stockfish.```
